File: c++/NeetCode150/4_group_anagrams.cpp
#include <algorithm>
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) 
    {
        unordered_map<string, vector<string>> anagram_groups;
        for (int i = 0; i < strs.size(); i++) {
            string sorted_s = strs[i];
            sort(sorted_s.begin(), sorted_s.end());
            anagram_groups[sorted_s].push_back(strs[i]);
        }

        vector<vector<string>> grouped_anagrams;
        for (auto& kv : anagram_groups) {
            grouped_anagrams.push_back(kv.second);
        }
        
        return grouped_anagrams;
    }
};

File: java/NeetCode150/4_group_anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> anagram_groups = new HashMap<String, List<String>>();
        for(int i = 0; i<strs.length; i++)
        {
            char[] c = strs[i].toCharArray();
            Arrays.sort(c);
            String sorted_s = new String(c);
            List<String> new_list;
            List<String> anagram_list = anagram_groups.getOrDefault(sorted_s, new ArrayList<>());
            anagram_list.add(strs[i]);
            anagram_groups.put(sorted_s, anagram_list);
        }
        return new ArrayList<>(anagram_groups.values());
    }
}
